### PR TITLE
Add the pytest-based integration testing jobs

### PR DESCRIPTION
**Pull request by paride**

### CODE REVIEW

**Comment by paride** [comment]
@TheRealFalcon @blackboxsw this is https://github.com/canonical/server-jenkins-jobs/pull/173 reopened against `main`. Apparently GitHub doesn't allow re-targeting a closed PRs, or at least closed PRs with a non-existing target branch, so I created a new one, also rebasing my branch.

**Review by TheRealFalcon — APPROVED** [review]
A couple questions/comments inline, but overall this looks good. Good thinking on the using the apt source for the daily builds rather than the git commit/tag.

I still don't know a ton about jenkins job builder. For the daily jobs, is there a way to schedule the jobs to run staggered on some kind of weekly schedule without having to define each individual job? If we eventually have 7 clouds (ec2, gce, azure, oracle, openstack, lxd containers, lxd vms) and 4 releases, that's 28 test runs we'll need to stagger.

Also, is it possible to for builds to be parameterized rather than creating so many jobs? It's certainly not a deal breaker, but it might be easier to manage.

File: jenkins-jobs/cloud-init/integration.yaml [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            lxc launch --quiet "$lxd_remote:{release}" "$container" --ephemeral
+            sleep 1
+            lxc exec "$container" -- cloud-init status --wait
+
+            if ((!proposed)); then
+                # Add the PPA pubkey.
+                kfile="/etc/apt/trusted.gpg.d/cloud-init-daily-ppa.gpg"
+                key="0x1ff0d8535ef7e719e5c81b9c083d06fbe4d304df"
+                keyserver="keyserver.ubuntu.com"
+                keyurl="https://${{keyserver}}/pks/lookup?op=get&search=${{key}}"
```
[the remark this answers is not in the document]
Comment (paride) [reply jenkins-jobs/cloud-init/integration.yaml:133]: Yes: they're escaped into a single brace.

Variables in single braces are substituted by JJB, and the double braces are the escape for this mechanism, producing a single brace in the actual job definition.

File: jenkins-jobs/cloud-init/integration.yaml [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                lxc exec "$container" -- sh -c "curl -sS '$keyurl' | gpg --dearmour --output $kfile"
```
[the remark this answers is not in the document]
Comment (paride) [reply jenkins-jobs/cloud-init/integration.yaml:134]: The single quote is within the `sh -c "..."` double quotes, so the variable gets expanded by the "outer" shell and its contents (a plain string) quoted within single quotes which are relevant only to `sh -c`. Same as doing:
```
$ x=x
$ sh -c "echo '$x'"
x
```
I think this is correct as it is.

File: jenkins-jobs/cloud-init/integration.yaml [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            # -vv --showlocals to show the full contents of variable in failed
+            # asserts, without ellipsis.
+            tox -e integration-tests -- -vv --showlocals tests/integration_tests
+
+
+- publisher:
+    name: cloud-init-archive-integration-logs
+    publishers:
+      - archive:
+          artifacts: 'cloud_init_test_logs'
```
[the remark this answers is not in the document]
Comment (paride) [reply jenkins-jobs/cloud-init/integration.yaml:182]: What about setting `CLOUD_INIT_COLLECT_LOGS=ON_ERROR`, so the logs only gets collected when they may contain something useful?

File: jenkins-jobs/cloud-init/integration.yaml [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    name: cloud-init-integration-pytest
+    platform:
+      - lxd_container
+      - lxd_vm
+      - ec2
+    release:
+      - xenial
+      - bionic
+      - focal
+      - hirsute
```
[the remark this answers is not in the document]
Comment (paride) [reply jenkins-jobs/cloud-init/integration.yaml:28]: And `impish` :)

File: jenkins-jobs/cloud-init/integration.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply jenkins-jobs/cloud-init/integration.yaml:182]: Good idea

[the remark this answers is not in the document]
Comment (paride) [reply jenkins-jobs/cloud-init/integration.yaml:182]: Done

**Comment by paride** [comment]
Hi @TheRealFalcon, I tweaked this a bit more and add a timed `@weekly` trigger to get the jobs running so we can start debugging them. If you're +1 I think we can merge this.

**Review by TheRealFalcon — COMMENTED** [review]
We still need to update the release list, otherwise I'm +1 to merge.

**Comment by paride** [comment]
Right, done!

**Review by TheRealFalcon — APPROVED** [review]

**Pull request merged**